agent/consul/acl.go

```diff
@@ -1265,77 +1272,83 @@ func (f *aclFilter) filterServiceNodes(nodes *structs.ServiceNodes) {
 		if f.allowNode(node.Node, &authzContext) && f.allowService(node.ServiceName, &authzContext) {
 			continue
 		}
+		removed = true
 		f.logger.Debug("dropping node from result due to ACLs", "node", structs.NodeNameString(node.Node, &node.EnterpriseMeta))
 		sn = append(sn[:i], sn[i+1:]...)
 		i--
 	}
 	*nodes = sn
+	return removed
 }
 
 // filterNodeServices is used to filter services on a given node base on ACLs.
-func (f *aclFilter) filterNodeServices(services **structs.NodeServices) {
+// Returns true if any elements were removed
+func (f *aclFilter) filterNodeServices(services **structs.NodeServices) bool {
 	if *services == nil {
-		return
+		return false
 	}
 
 	var authzContext acl.AuthorizerContext
 	(*services).Node.FillAuthzContext(&authzContext)
 	if !f.allowNode((*services).Node.Node, &authzContext) {
 		*services = nil
-		return
+		return true
 	}
 
+	var removed bool
 	for svcName, svc := range (*services).Services {
 		svc.FillAuthzContext(&authzContext)
 
 		if f.allowNode((*services).Node.Node, &authzContext) && f.allowService(svcName, &authzContext) {
 			continue
 		}
 		f.logger.Debug("dropping service from result due to ACLs", "service", svc.CompoundServiceID())
+		removed = true
 		delete((*services).Services, svcName)
 	}
+
+	return removed
 }
 
 // filterNodeServices is used to filter services on a given node base on ACLs.
-func (f *aclFilter) filterNodeServiceList(services **structs.NodeServiceList) {
-	if services == nil || *services == nil {
-		return
+// Returns true if any elements were removed.
+func (f *aclFilter) filterNodeServiceList(services *structs.NodeServiceList) bool {
+	if services.Node == nil {
+		return false
 	}
 
 	var authzContext acl.AuthorizerContext
-	(*services).Node.FillAuthzContext(&authzContext)
-	if !f.allowNode((*services).Node.Node, &authzContext) {
-		*services = nil
-		return
+	services.Node.FillAuthzContext(&authzContext)
+	if !f.allowNode(services.Node.Node, &authzContext) {
+		*services = structs.NodeServiceList{}
+		return true
 	}
 
-	svcs := (*services).Services
-	modified := false
+	var removed bool
+	svcs := services.Services
 	for i := 0; i < len(svcs); i++ {
 		svc := svcs[i]
 		svc.FillAuthzContext(&authzContext)
 
-		if f.allowNode((*services).Node.Node, &authzContext) && f.allowService(svc.Service, &authzContext) {
+		if f.allowService(svc.Service, &authzContext) {
```

Comment:
wait, why are we not filtering on the Node first as well ?

Reply:
I think it's safe to remove the check from this loop, because we're already checking node access on [L1322](https://github.com/hashicorp/consul/pull/11594/files/0714fc3260c9c96d69e194d69219d3feaa33507d#diff-e9bfbecd10c1931e8c8607866fcdc011f9701c9369fcbcefcc0187fc6638d0ddR1322) 🤞🏻